src/app/models/quiz_attempt.py

```diff
@@ -52,16 +95,32 @@ def top_categories(self):
         return self.graded_categories[:3]
 
     @classmethod
-    def get_by_user_id(cls, user_id):
+    def get_by_user_id(cls, user_id, quiz_type='fullform'):
         """
         Get first quiz attempt from a user id
         :param user_id: id of the user to get a quiz attempt for
+        :param quiz_type: type of quiz
         :rtype: QuizAttempt or None
         """
         if not user_id:
             raise ValueError('user_id must be provided')
+        if quiz_type not in cls.QUIZ_TYPES:
+            raise ValueError('quiz_type must be one of {}'.format(cls.QUIZ_TYPES))
+        if type(user_id) != int:
+            user_id = int(user_id)
+
+        return cls.build_key(user_id, quiz_type).get()
+
+    @classmethod
+    def get_all_attempts_for_user_id(cls, user_id):
+        """
+        Get all (up to three) attempts for the different quizzes.
+        :param user_id:
+        :return:
+        """
+        if not user_id:
+            raise ValueError('user_id must be provided')
         if type(user_id) != int:
             user_id = int(user_id)
-        attempts = cls.query(cls.user_id == user_id).fetch()
 
-        return attempts[0] if attempts else None
+        return filter(None, [cls.get_by_user_id(user_id, quiz_type) for quiz_type in cls.QUIZ_TYPES.keys()])
```

Comment:
They say you shouldn't use these map/filter/reduce functions much, but I like it in this case where it's only filtering by whether or not the items in the list are `None` or not.